### ISSUE TITLE
Training multiple workers with different train operations?

### ISSUE THREAD

**Issue by giladcohen** [uber/horovod#151]
I wish to implement synchronized training for MultiNet:

<img width="1107" alt="image" src="https://user-images.githubusercontent.com/20541403/35223844-89f4455e-ff8b-11e7-9587-ec4919ef3b37.png">
(paper in https://arxiv.org/pdf/1612.07695.pdf)

I want to use 3 workers training in parallel the 3 different heads (Classification/Detection/Segmentation). Each of my 3 nodes should have a specific train operation to calculate gradients for:
1) Node 0 will run classification_train_op to calculate the gradients of the Classification head + VGG16 backbone.
2) Node 1 will run detection_train_op to calculate the gradients of the Detection head + VGG16 backbone.
3) Node 2 will run segmentation_train_op to calculate the gradients of the Segmentation head + VGG16 backbone.

I want to extract all gradients after training (1),(2)(3) and average the VGG16 gradients. Can I employ Horovod for this task?

**Comment by alsrgv** [comment]
@giladcohen, does each of (1), (2), (3) fit on a single GPU with an appropriate batch size?  If so, you could manually place them there using `with tf.device('/gpu:X')` and use Horovod across these 3-GPU workers.

**Comment by giladcohen** [comment]
Yes, every head fits into a single GPU, but with different batch size. Is it problematic to run (1),(2),(3) with different batch_sizes?

I plan to write something like this:

enconder_out = convnet(input)
with tf.device('/gpu:0'):
    head0_loss = classification_head(enconder_out)
    head0_variables = tf.get_collection(<variables_in_backbone_and_head0>)
    head0_optimizer = tf.train.AdamOptimizer(0.001)
    head0_grads = tf.gradients(classification_loss, trainable_variables)
    
    # Average gradients using Horovod (influenced from your answer in issue #102):
    head0_averaged_gradients = []
    for grad, var in zip(head0_grads, head0_variables):
        if grad is not None:
            head0_averaged_gradients.append((hvd.allreduce(grad), var))
        else:
            head0_averaged_gradients.append((None, var))
    head0_apply_op = opt.apply_gradients(head0_averaged_gradients)

with tf.device('/gpu:1'):
    head1_loss = detection_head(enconder_out)
    head1_variables = tf.get_collection(<variables_in_backbone_and_head1>)
    head1_optimizer = ... # might be a different optimizer
    head1_grads = ...
    ...
    ...
    ....
    head1_apply_op = opt.apply_gradients(head1_averaged_gradients)

....
train_op = [head0_apply_op , head1_apply_op, head2_apply_op]

Is this a good implementation?

Also, after reading issue #49 , I understand that workers should update the batch norm parameters. I intend to update all the encoder's batch norm parameters using the task which utilizes the largest batch size, to get better statistics. The other two workers will not update the batch norm parameters in the encoder.
Also, workers must update the parameters in their correspoinding heads, if any batch norm is applied there, right?
Thanks.

**Comment by alsrgv** [comment]
Different batch sizes should not be a problem.

You probably don't need to explicitly do:
```python
    head0_grads = tf.gradients(classification_loss, head0_variables)
    
    # Average gradients using Horovod (influenced from your answer in issue #102):
    head0_averaged_gradients = []
    for grad, var in zip(head0_grads, head0_variables):
        if grad is not None:
            head0_averaged_gradients.append((hvd.allreduce(grad), var))
        else:
            head0_averaged_gradients.append((None, var))
    head0_apply_op = head0_optimizer.apply_gradients(head0_averaged_gradients)
```

Instead, you can just do regular `hvd.DistributedOptimizer`, but specify the device:

```python
head0_optimizer = hvd.DistributedOptimizer(head0_optimizer, device_dense='/gpu:0')
train_op = head0_optimizer.minimize(head0_loss)
```

Re: batch norm - which components will have batch norm layers - CNN encoder with VGG, heads?

**Comment by giladcohen** [comment]
Thank you @alsrgv.
Both the VGG encoder and the heads can have batch normalization layers.
Each worker needs to update the batch normalization parameters in its corresponding head, but how should I update the batch normalization parameters in the encoder?
Lets assume all the tasks employ the same batch_size.
If all workers shared the same input data, I could just choose one worker to update the batch norm parameters in the encoder. However, in my case every task is trained on different data. Therefore, I'd like to average also the encoder's batch norm updates amongst the tasks. How can I do that?

A follow up question - would your proposed solution work for tasks employing different batch_size?

Thanks.

**Comment by alsrgv** [comment]
I think if you share parameters among heads, BN parameters would get updated from all the sources of gradient updates.  In order to average them, you could adjust learning rate for VGG to be 1/3rd of typical learning rate, similar to https://github.com/tensorflow/tensorflow/issues/313.  Would that work?

**Comment by giladcohen** [comment]
Hi, not all the BN parameters are updated by the optimizers.

tf.layers.batch_normalization:
"""When training, the moving_mean and moving_variance need to be updated. By default the update ops are placed in tf.GraphKeys.UPDATE_OPS, so they need to be added as a dependency to the train_op."""

I can train many steps without updating the moving average/variance, but prior to running validation I need to somehow average the tf.GraphKeys.UPDATE_OPS corresponded to all the BN layers. Can it be done with horovod?

**Comment by alsrgv** [comment]
@giladcohen, what if you modify momentum of the batch norm to have 1/3 of "update power", i.e. 0.9967 instead of default 0.99, and then run updates from all three towers sequentially?  Should give you approximately the same answer without the code complexity.

**Comment by giladcohen** [comment]
Yes, setting 0.9967 instead of 0.99 could work if there was a shared parameter server that holds the current moving_mean and moving_average. Since every task (every GPU) updates its own moving values, how can they sync in horovod? The UPDATE_OPS are operations, not gradients.

**Comment by alsrgv** [comment]
@giladcohen, this way you would have per-GPU averaged batch norm parameters.  Keeping batch-norm parameters to bounds of GPU, rather than computing batch norm parameters across all servers was advocated in #49, as well as Facebook's [paper](https://research.fb.com/wp-content/uploads/2017/06/imagenet1kin1h5.pdf).

**Comment by giladcohen** [comment]
I'll try that. The issue can be closed.
Thanks @alsrgv!
